File: ltms/go/marxdata.go
// Marx Brothers puzzle data and rules.
// Converted from marxdata.lisp.
package ltms

// LoadMarxData loads the Marx Brothers puzzle constraints and rules.
func LoadMarxData(ltre *LTRE) {
	if ltre == nil {
		ltre = CurrentLTRE
	}
	old := CurrentLTRE
	CurrentLTRE = ltre
	defer func() { CurrentLTRE = old }()

	// Pairwise nogoods - attributes that can't be the same person
	RAssert([]interface{}{"PAIRWISE-NOGOOD", "PLAYS-PIANO", "PLAYS-HARP"}, "CONSTRAINT")
	RAssert([]interface{}{"PAIRWISE-NOGOOD", "PLAYS-PIANO", "SMOOTH-TALKER"}, "CONSTRAINT")
	RAssert([]interface{}{"PAIRWISE-NOGOOD", "PLAYS-HARP", "SMOOTH-TALKER"}, "CONSTRAINT")
	RAssert([]interface{}{"PAIRWISE-NOGOOD", "LIKES-GAMBLING", "LIKES-ANIMALS"}, "CONSTRAINT")
	RAssert([]interface{}{"PAIRWISE-NOGOOD", "SMOOTH-TALKER", "LIKES-GAMBLING"}, "CONSTRAINT")

	// Same entity constraints
	RAssert([]interface{}{"SAME-ENTITY", "LIKES-ANIMALS", "PLAYS-HARP"}, "CONSTRAINT")

	// Known facts
	RAssert([]interface{}{":NOT", []interface{}{"LIKES-ANIMALS", "GROUCHO"}}, "CONSTRAINT")
	RAssert([]interface{}{":NOT", []interface{}{"SMOOTH-TALKER", "HARPO"}}, "CONSTRAINT")
	RAssert([]interface{}{"PLAYS-PIANO", "CHICO"}, "CONSTRAINT")

	// Rule: pairwise-nogood means two attributes can't apply to the same object
	MakeRule([]interface{}{"PAIRWISE-NOGOOD", "?attribute1", "?attribute2"}, "TRUE",
		func(bindings map[string]interface{}, horNode *TmsNode) {
			attr1 := bindings["?attribute1"]
			attr2 := bindings["?attribute2"]

			MakeRule([]interface{}{attr1, "?obj"}, "TRUE",
				func(bindings2 map[string]interface{}, f1Node *TmsNode) {
					obj := bindings2["?obj"]

					MakeRule([]interface{}{attr2, obj}, "TRUE",
						func(bindings3 map[string]interface{}, f2Node *TmsNode) {
							RAssert([]interface{}{":NOT",
								[]interface{}{":AND",
									ViewNode(horNode),
									ViewNode(f1Node),
									ViewNode(f2Node)}},
								"PAIRWISE-NOGOOD-RULE")
						})
				})
		})

	// Rule: same-entity means if attribute1 applies to obj, so does attribute2
	MakeRule([]interface{}{"SAME-ENTITY", "?attribute1", "?attribute2"}, "TRUE",
		func(bindings map[string]interface{}, horNode *TmsNode) {
			attr1 := bindings["?attribute1"]
			attr2 := bindings["?attribute2"]

			MakeRule([]interface{}{attr1, "?obj"}, "TRUE",
				func(bindings2 map[string]interface{}, f1Node *TmsNode) {
					obj := bindings2["?obj"]
					RAssert([]interface{}{":IMPLIES",
						[]interface{}{":AND",
							ViewNode(horNode),
							ViewNode(f1Node)},
						[]interface{}{attr2, obj}},
						"SAME-ENTITY-RULE")
				})
		})

	RunRules(ltre)
}

File: ltms/java/MarxData.java
package ltms;

import java.util.*;

public class MarxData {

    public static void loadMarxData(LTRE ltre) {
        if (ltre == null) ltre = LTRE.currentLTRE;
        LTRE old = LTRE.currentLTRE;
        LTRE.currentLTRE = ltre;
        try {
            // Pairwise nogoods
            LTRE.assertFact(Arrays.asList("PAIRWISE-NOGOOD", "PLAYS-PIANO", "PLAYS-HARP"), "CONSTRAINT", ltre);
            LTRE.assertFact(Arrays.asList("PAIRWISE-NOGOOD", "PLAYS-PIANO", "SMOOTH-TALKER"), "CONSTRAINT", ltre);
            LTRE.assertFact(Arrays.asList("PAIRWISE-NOGOOD", "PLAYS-HARP", "SMOOTH-TALKER"), "CONSTRAINT", ltre);
            LTRE.assertFact(Arrays.asList("PAIRWISE-NOGOOD", "LIKES-GAMBLING", "LIKES-ANIMALS"), "CONSTRAINT", ltre);
            LTRE.assertFact(Arrays.asList("PAIRWISE-NOGOOD", "SMOOTH-TALKER", "LIKES-GAMBLING"), "CONSTRAINT", ltre);

            // Same entity
            LTRE.assertFact(Arrays.asList("SAME-ENTITY", "LIKES-ANIMALS", "PLAYS-HARP"), "CONSTRAINT", ltre);

            // Known facts
            LTRE.assertFact(Arrays.asList(":NOT", Arrays.asList("LIKES-ANIMALS", "GROUCHO")), "CONSTRAINT", ltre);
            LTRE.assertFact(Arrays.asList(":NOT", Arrays.asList("SMOOTH-TALKER", "HARPO")), "CONSTRAINT", ltre);
            LTRE.assertFact(Arrays.asList("PLAYS-PIANO", "CHICO"), "CONSTRAINT", ltre);

            // Rule: pairwise-nogood
            LtreRules.makeRule(Arrays.asList("PAIRWISE-NOGOOD", "?attribute1", "?attribute2"), "TRUE",
                (bindings, horNode) -> {
                    Object attr1 = bindings.get("?attribute1");
                    Object attr2 = bindings.get("?attribute2");

                    LtreRules.makeRule(Arrays.asList(attr1, "?obj"), "TRUE",
                        (bindings2, f1Node) -> {
                            Object obj = bindings2.get("?obj");

                            LtreRules.makeRule(Arrays.asList(attr2, obj), "TRUE",
                                (bindings3, f2Node) -> {
                                    LTRE.assertFact(Arrays.asList(":NOT",
                                        Arrays.asList(":AND",
                                            LTRE.viewNode(horNode),
                                            LTRE.viewNode(f1Node),
                                            LTRE.viewNode(f2Node))),
                                        "PAIRWISE-NOGOOD-RULE", LTRE.currentLTRE);
                                });
                        });
                });

            // Rule: same-entity
            LtreRules.makeRule(Arrays.asList("SAME-ENTITY", "?attribute1", "?attribute2"), "TRUE",
                (bindings, horNode) -> {
                    Object attr1 = bindings.get("?attribute1");
                    Object attr2 = bindings.get("?attribute2");

                    LtreRules.makeRule(Arrays.asList(attr1, "?obj"), "TRUE",
                        (bindings2, f1Node) -> {
                            Object obj = bindings2.get("?obj");
                            LTRE.assertFact(Arrays.asList(":IMPLIES",
                                Arrays.asList(":AND",
                                    LTRE.viewNode(horNode),
                                    LTRE.viewNode(f1Node)),
                                Arrays.asList(attr2, obj)),
                                "SAME-ENTITY-RULE", LTRE.currentLTRE);
                        });
                });

            LtreRules.runRules(ltre);
        } finally {
            LTRE.currentLTRE = old;
        }
    }
}
